app/services/fixity_checker.rb

```diff
@@ -0,0 +1,117 @@
+# FixityChecker.new(asset).check
+# FixityChecker.new(asset).prune_checks
+# This first version of the fixity checker only checks
+# for actual discrepancies between already-recorded
+# checksums and calculated ones.
+#
+# This means:
+# Assets with nil files are ignored.
+# Files with nil checksums are ignored.
+
+class FixityChecker
+  attr_reader :asset
+
+  NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP = 5
+  CHECK_CYCLE_LENGTH = 7
+
+
+  # @param work [Work] Work object, it's members will be put into a zip
+  # @param callback [proc], proc taking keyword arguments progress_i: and progress_total:, can
+  #   be used to update a progress UI.
+  def initialize(asset)
+    raise ArgumentError.new("Please pass in an Asset.") unless asset.is_a? Asset
+    @asset = asset
+  end
+
+  #FixityChecker.new(asset).check
+  def check(sieve_integer = 0)
+    return unless sift(sieve_integer)
+    return nil if @asset.nil?
+    return nil if @asset.file.nil?
+    the_expected_checksum = expected_checksum
+    # Note: @asset.file.sha512 can be nil on e.g.
+    # items imported with DISABLE_BYTESTREAM_IMPORT=true
+    # For now, we are ignoring these.
+    return nil if the_expected_checksum.nil?
+
+    new_check = FixityCheck.new(asset: @asset)
+    new_check.expected_result = the_expected_checksum
+    new_check.checked_uri = @asset.file.url
+    new_check.actual_result = actual_checksum
+    new_check.passed= (the_expected_checksum==actual_checksum)
+    new_check.save!
+  end
+
+  #FixityChecker.new(asset).prune_checks
+  # After pruning, for each asset - URI combination, you will be left with:
+  # up to NUMBER_OF_RECENT_PASSED_CHECKS_TO_KEEP recent passed checks,
+  # plus all the failed checks,
+  # plus one initial check (to remember the file that was intially uploaded.)
+  def prune_checks(sieve_integer = 0)
+    return unless sift(sieve_integer)
```

Comment:
I don't understand what's going on with `sift`. If later `checks_its_okay_to_delete` returns an empty array, then none will be deleted. What is the extra `sift` for? Ah, I see, it's about trying to create a 'cycle' -- I am not sure this is doing it a good way, I need you to talk me through what it's doing.

Reply:
Moved the sift function out to the rake task where it makes more sense to put it, and also rewrote it and commented it a bit better.